src/main/scala/io/iohk/ethereum/BootstrapDownload.scala

```diff
@@ -0,0 +1,135 @@
+package io.iohk.ethereum
+
+
+import java.io.{File, FileInputStream, FileOutputStream}
+import java.net.URL
+import java.nio.file._
+import java.security.{DigestInputStream, MessageDigest}
+import java.util.zip.ZipInputStream
+
+import io.iohk.ethereum.utils.Logger
+
+
+/**
+  * A facility to
+  * - check the download location for a minimum amount of free space
+  * - download a zip from a URL and generate MD5 checksum
+  * - check the checksum
+  * - clean files out of given location
+  * - unzip to a given location
+  */
+object BootstrapDownload extends Logger {
+
+  val bufferSize = 4 * 1024
+  val leveldbFolderName = "leveldb"
+
+  private def assertAndLog(cond: Boolean, msg: String): Unit = {
+    if(!cond) log.info(msg)
+    assert(cond, msg)
+  }
+
+  def cleanOutFolder(pathToDownloadTo: Path): Unit = {
+    val leveldbFolder = pathToDownloadTo.toFile
+    assertAndLog(leveldbFolder.isDirectory, s"${pathToDownloadTo} must be a folder.")
+    assertAndLog(leveldbFolder.getName == leveldbFolderName, s"${pathToDownloadTo} must end in a folder named $leveldbFolderName")
+    leveldbFolder.listFiles().foreach(_.delete())
+  }
+
+  def downloadFile(urlToDownloadFrom: String, outFile: File): String = {
+
+    val md5 = MessageDigest.getInstance("MD5")
+    val is = new URL(urlToDownloadFrom).openStream()
+    try {
+      val dis = new DigestInputStream(is, md5)
+      try {
+        val out = new FileOutputStream(outFile)
+        try {
+          val buffer = new Array[Byte](bufferSize)
+          Stream.continually(dis.read(buffer)).takeWhile(_ != -1).foreach(out.write(buffer, 0, _))
+        } finally (out.close())
+        md5.digest.map("%02x".format(_)).mkString
+      } finally(dis.close())
+    } finally(is.close())
```

Comment:
How about using loan pattern to close the resources (akin to Java's ARM)? It might look something like:

```
def using[T <: { def close() }, U](resource: T)(body: T => U): U = {
  try {
    body(resource)
  } finally {
    resource.close()
  }
}
```

Then:

```
using(new URL(urlToDownloadFrom).openStream()) { is =>
  readStream(is)
}
```

Reply:
Also, we don't need as many `finally`s, `dis` will close `is`.

Reply:
I did think about using another pattern for resource closing, I like it cos it's neat, but it's very debatably clearer. A quick search of the codebase shows we use 'try'/'finally'. That might be a post 1.0 task. 

I just checked and the Digest constructor is pretty safe so we can put them together. I had them together but I separated them wondering if 

`val dis = new DigestInputStream(new URL(urlToDownloadFrom).openStream(), md5)`

... DigestInputStream throws in the constructor is that a resource leak? In this case it's not.